cpp/src/arrow/array/util.cc

```diff
@@ -669,11 +671,29 @@ class RepeatedArrayFactory {
   enable_if_base_binary<T, Status> Visit(const T&) {
     const std::shared_ptr<Buffer>& value = scalar<T>().value;
     std::shared_ptr<Buffer> values_buffer, offsets_buffer;
-    RETURN_NOT_OK(CreateBufferOf(value->data(), value->size(), &values_buffer));
     auto size = static_cast<typename T::offset_type>(value->size());
+
+    typename T::offset_type length_casted;
+    if (length_ < 0) {
+      return Status::Invalid("length cannot be negative: " + std::to_string(length_));
```

Comment:
I think the negativity check can be moved to the top level `MakeArrayFromScalar` since it applies to all types.